zkvm/src/patricia.rs

```diff
@@ -0,0 +1,520 @@
+use merlin::Transcript;
+use std::cmp::min;
+
+use crate::errors::VMError;
+
+/// Patricia tree for constructing Merkle proofs of
+/// inclusion.
+pub struct PatriciaTree {
+    root: Node,
+    size: usize,
+}
+
+/// Specifies a neighbor in the proof of inclusion.
+#[derive(Clone, Debug)]
+pub enum PatriciaNeighbor {
+    /// Left
+    Left([u8; 32]),
+    /// Right
+    Right([u8; 32]),
+}
+
+struct Node {
+    /// Data held by this node
+    item: Option<PatriciaItem>,
+    /// Hash of the data in this node (if a leaf),
+    /// or of the children concatenated
+    hash: [u8; 32],
+    /// Left (bit 0) and right (bit 1) nodes
+    children: Option<(Box<Node>, Box<Node>)>,
+}
+
+#[derive(Clone)]
+struct PatriciaItem {
+    /// Key contained at this item: with the appropriate
+    /// bits included, represents the data represented
+    /// by this path of the Patricia tree.
+    key: &'static [u8],
```

Comment:
`'static` is not going to work with actual data. Since all our use cases are for 32-byte keys, we can hardcode the key as `[u8;32]`.